cpp/src/parquet/chunker_internal_test.cc

```diff
@@ -0,0 +1,1397 @@
+
+  // with norm_factor=-1 the difference between min and max chunk size must be
+  // at least 8
+  ASSERT_THROW(ContentDefinedChunker(li, 0, 7, -1), ParquetException);
+  ASSERT_NO_THROW(ContentDefinedChunker(li, 0, 8, -1));
+  ASSERT_NO_THROW(ContentDefinedChunker(li, 0, 16, -1));
+
+  // test the norm_factor extremes
+  ASSERT_THROW(ContentDefinedChunker(li, 0, 0, -68), ParquetException);
+  ASSERT_NO_THROW(ContentDefinedChunker(li, 0, 1, -67));
+  ASSERT_THROW(ContentDefinedChunker(li, 0, std::numeric_limits<int64_t>::max(), 59),
+               ParquetException);
+  ASSERT_NO_THROW(ContentDefinedChunker(li, 0, std::numeric_limits<int64_t>::max(), 58));
+}
+
+class TestCDCSingleRowGroup : public ::testing::TestWithParam<CaseConfig> {
+ protected:
+  // Column random table parts for testing
+  std::shared_ptr<Field> field_;
+  std::shared_ptr<Table> part1_, part2_, part3_, part4_, part5_, part6_, part7_;
+
+  void SetUp() override {
+    const auto& param = GetParam();
+    auto field_ = ::arrow::field("f0", param.dtype, param.is_nullable);
+    auto schema = ::arrow::schema({field_});
+
+    auto part_length = kPartSize / param.bytes_per_record;
+    auto edit_length = kEditSize / param.bytes_per_record;
+    ASSERT_OK_AND_ASSIGN(part1_, GenerateTable(schema, part_length, 0));
+    ASSERT_OK_AND_ASSIGN(part2_, GenerateTable(schema, edit_length, 1));
+    ASSERT_OK_AND_ASSIGN(part3_, GenerateTable(schema, part_length, part_length));
+    ASSERT_OK_AND_ASSIGN(part4_, GenerateTable(schema, edit_length, 2));
+    ASSERT_OK_AND_ASSIGN(part5_, GenerateTable(schema, part_length, 2 * part_length));
+    ASSERT_OK_AND_ASSIGN(part6_, GenerateTable(schema, edit_length, 3));
+    ASSERT_OK_AND_ASSIGN(part7_, GenerateTable(schema, edit_length, 4));
+  }
+};
+
+TEST_P(TestCDCSingleRowGroup, DeleteOnce) {
+  const auto& param = GetParam();
+
+  ASSERT_OK_AND_ASSIGN(auto base, ConcatAndCombine({part1_, part2_, part3_}));
+  ASSERT_OK_AND_ASSIGN(auto modified, ConcatAndCombine({part1_, part3_}));
+  ASSERT_FALSE(base->Equals(*modified));
+
+  for (bool enable_dictionary : {false, true}) {
+    ASSERT_OK_AND_ASSIGN(auto base_info, WriteAndGetParquetInfo(
+                                             base, kMinChunkSize, kMaxChunkSize,
+                                             enable_dictionary, param.data_page_version));
+    ASSERT_OK_AND_ASSIGN(
+        auto modified_info,
+        WriteAndGetParquetInfo(modified, kMinChunkSize, kMaxChunkSize, enable_dictionary,
+                               param.data_page_version));
+
+    // assert that there is only one row group
+    ASSERT_EQ(base_info.size(), 1);
+    ASSERT_EQ(modified_info.size(), 1);
+    AssertChunkSizes(param.dtype, base_info.front(), modified_info.front(),
+                     param.is_nullable, enable_dictionary, kMinChunkSize, kMaxChunkSize);
+
+    auto edit_length = part2_->num_rows();
+    if (::arrow::is_list_like(param.dtype->id())) {
+      edit_length += 1;
+    }
```

Comment:
Please add a comment explaining this.

Reply:
I made it more robust adding the number of `null` / `empty list` elements to the number of values.